### ISSUE TITLE
DL2 loss function generates infinite constraint loss

### ISSUE THREAD

**Issue by jh4n23** [vehicle-lang/vehicle#1198]
When attempting to train a Fashion MNIST classifier with robustness, the DL2 loss function generates infinite constraint loss. This is especially strange, as the compiled loss function is identical to the Vehicle Loss function, though training with this does __not__ cause infinite constraint loss.

See: [DL2Loss_compiled.txt](https://github.com/user-attachments/files/29884842/DL2Loss_compiled.txt) vs. [VehicleLoss_compiled.txt](https://github.com/user-attachments/files/29884852/VehicleLoss_compiled.txt).

Specification (using property "robust"): [mnist-robustness.txt](https://github.com/user-attachments/files/29884914/mnist-robustness.txt)

**Comment by MatthewDaggitt** [comment]
Will try and fix this when after I get the forcing refactor through and that generated code actually collapses down to a reasonable size, instead of being endless copies of the same tensor over and over again.

**Comment by MatthewDaggitt** [comment]
Okay I have got to the bottom of this. 

1. The DL2 logic has no good notion of negation as `True` is mapped to [0,0] and `False` is mapped to (0, infinity) so there is no way to map from one to the other bijectively. We cheat by adding `not x ---> 1/x` but as the below shows it doesn't really work.
2. Vehicle translates `forall x . P x` into `not (exists x . not (P x))`, which allows us to only have a single notion of searching for an existential witness in the Python backend. 
3. The problem with the robustness is spec is then the search succeeds in finding a witness (i.e. an input that flips the classification) and returns `0` representing true. But DL2 doesn't tell us _how true_ this witness is. So when we negate it by taking `1/x` we get `infinity` which produces no gradients.

So either we have to give up on DL2 or I have to work out how to alter the Vehicle backend not to convert `exists` into `forall`s...
